# **ENGG680 - Introduction to Digital Engineering**
## *Lab Assignment 4: ML Applications - Vehicle Motion State Classification*

## Preliminary: Certificate of Work


*We, the undersigned, certify that this is our own work, which has been done expressly for this course, either without the assistance of any other party or where appropriate we have acknowledged the work of others. Further, we have read and understood the section in the university calendar on plagiarism/cheating/other academic misconduct and we are aware of the implications thereof. We request that the total mark for this assignment be distributed as follows among group members:*



```
# This is formatted as code
```

|          | First Name | Last Name | Signature (Full Name, Date) | Hours | Contribution % |
|----------|------------|-----------|-----------------------------|-------|----------------|
| Member 1: | Kapildev  | Aryal | Kapildev Aryal | 2.5 Hours | 25% Contribution |
| Member 2: | Pawan | Bhusal | Pawan Bhusal| 2.5 Hours | 25% Contribution |
| Member 3: |Manesh | Pandit Chhetry | Maneshpanditchhetry |2.5 Hours | 25% Contribution |
| Member 4: | Samit | Shrestha | Samitshrestha | 2.5 Hours |25% Contribution |


---

## Section 1 - Motion Classification Using *All Features* Derived From Inertial Sensor Data (33 Marks)

---

### 1.1 Assignment datasets import (6 Marks)

---

In [1]:
# some library imports
import pandas as pd
# Import google drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


First load the following datasets into dataframes.
* *UrbanNav-HK_Whampoa-20210521_sensors.csv* to dataframe variable named *cv_set*. This dataset is to be used for model training and hyper parameter tuning.
* *UrbanNav-HK_TST-20210517_sensors.csv* to dataframe variable named *test_set1*. This is first independent test dataset.
* *2020-03-14-16-45-35.csv* to dataframe variable named *test_set2*. This is the second independent test dataset.

In [2]:
# (2 marks)
# TODO: Load cross validation/training dataset (UrbanNav-HK_Whampoa-20210521_sensors.csv) to 'cv_set' dataframe
cv_set = pd.read_csv('/content/drive/MyDrive/engg680_2024_fall/Lab_Assignment4/UrbanNav-HK_Whampoa-20210521_sensors.csv')

# TODO: Load UrbanNav-HK_TST-20210517_sensors.csv to 'test_set1' dataframe
test_set1 = pd.read_csv('/content/drive/MyDrive/engg680_2024_fall/Lab_Assignment4/UrbanNav-HK_TST-20210517_sensors.csv')

# TODO: Load 2020-03-14-16-45-35.csv to 'test_set2' dataframe
test_set2 = pd.read_csv('/content/drive/MyDrive/engg680_2024_fall/Lab_Assignment4/2020-03-14-16-45-35.csv')

For the *cv_set*, look at the dataframe head and info, and print the list of all columns. The *test_set1* and *test_set2* have the same schema as *cv_set*, however, have different number of rows (samples).

In [3]:
# TODO: Display first two rows of cv_set
# (1 mark)
cv_set.head(2)

ax_sum    ay_sum    az_sum    gx_sum    gy_sum    gz_sum     a_sum  \
0 -0.003929 -0.093081  9.789349 -0.002649  0.001406 -0.001391  9.797418   
1 -0.000966 -0.092417  9.789095 -0.001806  0.000103 -0.001943  9.798815   

      g_sum        ax_var        ay_var  ...     g_min    ax_max    ay_max  \
0  0.022554  8.046094e-07  1.319665e-07  ...  0.000003  0.001496  0.000544   
1  0.021731  1.032619e-06  1.073067e-07  ...  0.000006  0.001865  0.000615   

     az_max    gx_max    gy_max    gz_max     a_max     g_max  motion_state  
0  0.024992  0.000110  0.000075  0.000078  0.024997  0.000187          rest  
1  0.024828  0.000124  0.000098  0.000090  0.024829  0.000144          rest  

[2 rows x 33 columns]

In [4]:
# TODO: Load the info method of cv_set
# (1 mark)
cv_set.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1534 entries, 0 to 1533
Data columns (total 33 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   ax_sum        1534 non-null   float64
 1   ay_sum        1534 non-null   float64
 2   az_sum        1534 non-null   float64
 3   gx_sum        1534 non-null   float64
 4   gy_sum        1534 non-null   float64
 5   gz_sum        1534 non-null   float64
 6   a_sum         1534 non-null   float64
 7   g_sum         1534 non-null   float64
 8   ax_var        1534 non-null   float64
 9   ay_var        1534 non-null   float64
 10  az_var        1534 non-null   float64
 11  gx_var        1534 non-null   float64
 12  gy_var        1534 non-null   float64
 13  gz_var        1534 non-null   float64
 14  a_var         1534 non-null   float64
 15  g_var         1534 non-null   float64
 16  ax_min        1534 non-null   float64
 17  ay_min        1534 non-null   float64
 18  az_min        1534 non-null 

In [5]:
# TODO: Display the number of rows and columns  of all three datasets
# (1 mark)
print(f"'cv_set' has {cv_set.shape[0]} rows and {cv_set.shape[1]} columns")
print(f"'test_set1' has {test_set1.shape[0]} rows and {test_set1.shape[1]} columns")
print(f"'test_set2' has {test_set2.shape[0]} rows and {test_set2.shape[1]} columns")

'cv_set' has 1534 rows and 33 columns
'test_set1' has 783 rows and 33 columns
'test_set2' has 297 rows and 33 columns


In [6]:
# TODO: Display all columns in cv_set
# (1 mark)
print(cv_set.columns)

Index(['ax_sum', 'ay_sum', 'az_sum', 'gx_sum', 'gy_sum', 'gz_sum', 'a_sum',
       'g_sum', 'ax_var', 'ay_var', 'az_var', 'gx_var', 'gy_var', 'gz_var',
       'a_var', 'g_var', 'ax_min', 'ay_min', 'az_min', 'gx_min', 'gy_min',
       'gz_min', 'a_min', 'g_min', 'ax_max', 'ay_max', 'az_max', 'gx_max',
       'gy_max', 'gz_max', 'a_max', 'g_max', 'motion_state'],
      dtype='object')


In all three dataframes, target vector is stored in the *motion_state* column. All other columns are features.

___

### 1.2 - Grid Search with Cross Validation (18 Marks)

___

#### 1.2.1 - Extract all features and target from cv_set

___

From the *cv_set*, load the target vector stored in the *motion_state* column into a new variable *y*

In [7]:
# TODO - Load target vector into y
# (1 mark)
y = cv_set['motion_state']

From *cv_set*, load all features, that is all columns except for the target vector, in a new dataframe variable *X*

In [8]:
# TODO - Load all features from cv_set to X
# (2 marks)
X = cv_set.drop('motion_state', axis=1)
# Display the first two rows of X
X.head(2)

ax_sum    ay_sum    az_sum    gx_sum    gy_sum    gz_sum     a_sum  \
0 -0.003929 -0.093081  9.789349 -0.002649  0.001406 -0.001391  9.797418   
1 -0.000966 -0.092417  9.789095 -0.001806  0.000103 -0.001943  9.798815   

      g_sum        ax_var        ay_var  ...     a_min     g_min    ax_max  \
0  0.022554  8.046094e-07  1.319665e-07  ...  0.023871  0.000003  0.001496   
1  0.021731  1.032619e-06  1.073067e-07  ...  0.024153  0.000006  0.001865   

     ay_max    az_max    gx_max    gy_max    gz_max     a_max     g_max  
0  0.000544  0.024992  0.000110  0.000075  0.000078  0.024997  0.000187  
1  0.000615  0.024828  0.000124  0.000098  0.000090  0.024829  0.000144  

[2 rows x 32 columns]

___

#### 1.2.2 - Configure and Perform Grid Search with Cross-Validation

___

In [9]:
from sklearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.preprocessing import StandardScaler, RobustScaler, MinMaxScaler, Normalizer
from sklearn.model_selection import GridSearchCV

The dataset *(X, y)*, must be split into training/validation *(X_train, y_train)* and testing *(X_test, y_test)* sets. Use the `train_test_split` function from sklearn to create the split, where  **30%** of data is used for testing. Set the `random_state=42`, and `stratify=y`.  

In [10]:
# TODO - Perform test/train split
# (1 mark)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42, stratify=y)

Initialize a `Pipeline`, consisting of a *scaler* and a *classifer*, set the default values to *StandardScaler()*, and *SVC()*, respectively.

In [11]:
# TODO - Initialize Pipeline into variable pipe
# (2 marks)
pipe = Pipeline([
    ('scaler', StandardScaler()),
    ('classifier', SVC())
])

Define the grid search configuration (*param_grid*) for training and hyper parameter tuning of the *support vector classifier*. Tune, the hyper-parameter *C*, over the interval `C=[0.01, 0.1, 1 ,10,100]`, *gamma*, over the interval `gamma=[0.001, 0.01, 0.1,1]`, and use the *radial basis function* as the kernel. For the *scaler* parameter, test with the `StandardScaler`, `RobustScaler`, and `None` options.

In [12]:
# TODO - Setup grid set
# (4 marks)

param_grid = {
    'scaler': [StandardScaler(), RobustScaler(), None],
    'classifier__C': [0.01, 0.1, 1, 10, 100],
    'classifier__gamma': [0.001, 0.01, 0.1, 1],
    'classifier__kernel': ['rbf']
}

In [13]:
# Perform the grid search to train model and tune hyper parameters
grid = GridSearchCV(pipe, param_grid, cv=5, scoring='accuracy', return_train_score=True)
grid.fit(X_train, y_train)

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('scaler', StandardScaler()),
                                       ('classifier', SVC())]),
             param_grid={'classifier__C': [0.01, 0.1, 1, 10, 100],
                         'classifier__gamma': [0.001, 0.01, 0.1, 1],
                         'classifier__kernel': ['rbf'],
                         'scaler': [StandardScaler(), RobustScaler(), None]},
             return_train_score=True, scoring='accuracy')

___

#### 1.2.3 - Visualize Grid Search Results

___

Print the results from the grid search. The following outputs must be printed.
* Best parameter set
* Best cross-validation train score
* Best cross-validation test score
* Test set accuracy

In [14]:
# TODO: Print the results from the grid search
# (2 marks)

print("Best params:\n{}\n".format(grid.best_params_))
# Calculate the mean train score for the best parameter set
import numpy as np
best_index = grid.best_index_
train_score = np.mean(grid.cv_results_['mean_train_score'][best_index])
# Print the best cross-validation train score
print("Best cross-validation train score: {:.2f}".format(train_score))
print("Best cross-validation validation score: {:.2f}".format(grid.best_score_))
print("Test-set score: {:.2f}".format(grid.score(X_test, y_test)))

Best params:
{'classifier__C': 100, 'classifier__gamma': 0.01, 'classifier__kernel': 'rbf', 'scaler': RobustScaler()}

Best cross-validation train score: 0.98
Best cross-validation validation score: 0.95
Test-set score: 0.96


Display the confusion matrix and print the classification report for the trained model. Use the *(X_test, y_test)* dataset. **Hint** - You can use your code from assignment 3.

In [15]:
# TODO - Display confusion matrix and print classification report for (X_test, y_test)
# (2 marks)
from sklearn.metrics import confusion_matrix, classification_report
# Make predictions on the test set
y_pred = grid.predict(X_test)
# Display the confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:\n", conf_matrix)
# Print the classification report
class_report = classification_report(y_test, y_pred)
print("Classification Report:\n", class_report)

Confusion Matrix:
 [[153   2   0]
 [ 10 192   5]
 [  0   3  96]]
Classification Report:
               precision    recall  f1-score   support

        rest       0.94      0.99      0.96       155
    straight       0.97      0.93      0.95       207
        turn       0.95      0.97      0.96        99

    accuracy                           0.96       461
   macro avg       0.95      0.96      0.96       461
weighted avg       0.96      0.96      0.96       461



___

#### 1.2.4 - Questions

1. Comment on the fit of the model to the dataset? (2 marks)
2. In the grid search, are the intervals used for *C* and *gamma* hyperparameters suitable? Why or why not? (2 marks)

#### 1.2.4 - Answers
1. The model performs well, achieving high accuracy, precision, and recall, which indicates it effectively distinguishes between the classes.

2.  Yes, the intervals for C and gamma are suitable as they cover a wide range, ensuring a comprehensive search for optimal hyperparameters.

___

### 1.3 - Model Evaluation on Test Datasets 1 and 2 (9 Marks)

___

#### 1.3.1 - Extract all features and targets from test datasets

Next, the model trained in 1.2 will be tested on independent datasets, that is *test_set1* and *test_set2*. <br>
The test_sets must be seperated into features (Xs) and targets (ys).

In [16]:
# TODO - Load the target vector from test_set1 into y1, and all features from test_set1 into X1
# (1 marks)
y1 = test_set1[ 'motion_state']
X1 = test_set1.drop('motion_state', axis=1)
# Display the first two rows of X1
X1.head(2)

ax_sum    ay_sum    az_sum    gx_sum    gy_sum    gz_sum     a_sum  \
0  0.264465  0.089904  9.786930 -0.002305  0.000218 -0.001143  9.801661   
1  0.264300  0.088882  9.788127 -0.001865 -0.000523 -0.001999  9.802481   

      g_sum    ax_var        ay_var  ...     a_min     g_min    ax_max  \
0  0.025634  0.000001  7.749919e-08  ...  0.024056  0.000003  0.002506   
1  0.025218  0.000001  9.243000e-08  ...  0.024084  0.000007  0.002427   

     ay_max    az_max    gx_max    gy_max    gz_max     a_max     g_max  
0  0.000845  0.024884  0.000155  0.000078  0.000069  0.024886  0.000170  
1  0.000949  0.024881  0.000201  0.000095  0.000076  0.024883  0.000201  

[2 rows x 32 columns]

In [17]:
# TODO - Load the target vector from test_set2 into y2, and all features from test_set2 into X2
y2 = test_set2['motion_state']
X2 = test_set2.drop('motion_state', axis=1)
X2.head(2)

ax_sum    ay_sum    az_sum    gx_sum    gy_sum    gz_sum     a_sum  \
0 -0.046400  0.305907  9.791451 -0.002184  0.001311 -0.001901  9.800683   
1 -0.050073  0.311947  9.785619 -0.003369  0.001898 -0.002051  9.797741   

      g_sum    ax_var        ay_var  ...     a_min     g_min    ax_max  \
0  0.012437  0.000002  3.020959e-07  ...  0.048406  0.000007  0.002575   
1  0.013805  0.000003  6.859937e-07  ...  0.047696  0.000013  0.003880   

     ay_max    az_max    gx_max    gy_max    gz_max     a_max     g_max  
0  0.002993  0.049701  0.000094  0.000115  0.000080  0.049764  0.000142  
1  0.004106  0.050477  0.000140  0.000140  0.000083  0.050763  0.000151  

[2 rows x 32 columns]

____

#### 1.3.2 - Visualize Results

For each test dataset, print the test set accuracy score.

In [18]:
# TODO - Print test score accuracies for test_set1 and test_set2
# (1 marks)
test_set1_score = grid.score(X1, y1)
test_set2_score = grid.score(X2, y2)

print("test_set1 score: {:.2f}".format(test_set1_score))
print("test_set2 score: {:.2f}".format(test_set2_score))

test_set1 score: 0.92
test_set2 score: 0.25


Display the confusion matrix and print the classification for test_set 1 *(X1, y1)*

In [19]:
# TODO - Display confusion matrix and print classification report for (X1, y1)
# (1 marks)

# Make predictions on the test set 1
y1_pred = grid.predict(X1)
# Display the confusion matrix
conf_matrix1 = confusion_matrix(y1, y1_pred)
print("Confusion Matrix for test_set1:\n", conf_matrix1)
# Print the classification report
class_report1 = classification_report(y1, y1_pred)
print("Classification Report for test_set1:\n", class_report1)

Confusion Matrix for test_set1:
 [[200  31   0]
 [  7 387  17]
 [  0   7 134]]
Classification Report for test_set1:
               precision    recall  f1-score   support

        rest       0.97      0.87      0.91       231
    straight       0.91      0.94      0.93       411
        turn       0.89      0.95      0.92       141

    accuracy                           0.92       783
   macro avg       0.92      0.92      0.92       783
weighted avg       0.92      0.92      0.92       783



Display the confusion matrix and print the classification for test_set 2 *(X2, y2)*

In [20]:
# TODO - Display confusion matrix and print classification report for (X2, y2)

# Make predictions on the test set 2
y2_pred = grid.predict(X2)
# Display the confusion matrix
conf_matrix2 = confusion_matrix(y2, y2_pred)
print("Confusion Matrix for test_set2:\n", conf_matrix2)
# Print the classification report
class_report2 = classification_report(y2, y2_pred)
print("Classification Report for test_set2:\n", class_report2)


Confusion Matrix for test_set2:
 [[  0   0  16]
 [  0   0 206]
 [  0   0  75]]
Classification Report for test_set2:
               precision    recall  f1-score   support

        rest       0.00      0.00      0.00        16
    straight       0.00      0.00      0.00       206
        turn       0.25      1.00      0.40        75

    accuracy                           0.25       297
   macro avg       0.08      0.33      0.13       297
weighted avg       0.06      0.25      0.10       297



/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


___

#### 1.3.3 - Questions

**Based on results so far**
1. Comment on the model's performances on the datasets *(X_test, y_test)*, *(X1, y1)*, *(X2, y2)*? A significant difference in performance should be observed, what is this an example of? (3 marks)
1. Provide and explain potential causes for the observed behaviour in performance? (3 marks)

#### 1.3.3 - Answers
1. (X_test, y_test): The model performs very well with high accuracy and balanced metrics across all classes.

(X1, y1): The model also performs well with a high accuracy score and balanced metrics, indicating good generalization to this dataset.

(X2, y2): The model performs poorly with a low accuracy score and fails to predict certain classes accurately.

The significant difference in performance between (X_test, y_test)/(X1, y1) and (X2, y2) is an example of dataset shift or distribution shift. This occurs when the distribution of the training data differs significantly from the distribution of the test data.


2. Potential Causes for Observed Behavior:

a) Feature Distribution Differences: The features in test_set2 might have different distributions compared to the training set, leading to poor performance.

b) Class Imbalance: There might be an imbalance in the classes within test_set2 that the model wasn't adequately trained on.

c) Overfitting to Training Data: The model might be overfitting to the training data, capturing noise rather than the underlying patterns, and thus not generalizing well to test_set2.

___

## Section 2 - Classification Using *Selected Features* Derived From Inertial Sensor Data (12 Marks)

In this second section, an *SVC* model will be trained and tuned using grid search, however, only *selected features* will be used.

___

### 2.1 Grid Search with Cross Validation (5 Marks)

___

#### 2.1.1 Extract *selected* features and target from *cv_set*

From the *cv_set* dataframe, extract the selected features, listed below, and load them in a new dataframe *X*.
* *ax_var* - variance of accelerometer readings in x
* *ay_var* - variance of accelerometer readings in y
* *az_sum* - sum of accelerometer z readings
* *a_sum* - sum of the norm of accelerometer x, y, z readings
* *gz_var* - variance of gyroscope reading in z

In [21]:
# TODO - Load selected features from cv_set to X
# (1 marks)
selected_features = ['ax_var', 'ay_var', 'az_sum', 'a_sum', 'gz_var']
X = cv_set[selected_features]
# Display the first two rows of X
X.head(2)

ax_var        ay_var    az_sum     a_sum        gz_var
0  8.046094e-07  1.319665e-07  9.789349  9.797418  7.446916e-10
1  1.032619e-06  1.073067e-07  9.789095  9.798815  6.983481e-10

In [22]:
# TODO - Load motion_state column (target vector) into y
y = cv_set['motion_state']

___

#### 2.1.2 - Configure and Perform Grid Search with Cross-Validation

Follow the steps from **1.2.2** and perform model training and tuning using the selected feature set.

In [23]:
# TODO - Perform model training and tuning using `GridSearchCV`
# Split the data using the selected features
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42, stratify=y)
# Setup grid set
param_grid = {
    'scaler': [StandardScaler(), RobustScaler(), None],
    'classifier__C': [0.01, 0.1, 1, 10, 100],
    'classifier__gamma': [0.001, 0.01, 0.1, 1],
    'classifier__kernel': ['rbf']
}

# Perform the grid search to train model and tune hyper parameters
grid = GridSearchCV(pipe, param_grid, cv=5, scoring='accuracy', return_train_score=True)
grid.fit(X_train, y_train)


GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('scaler', StandardScaler()),
                                       ('classifier', SVC())]),
             param_grid={'classifier__C': [0.01, 0.1, 1, 10, 100],
                         'classifier__gamma': [0.001, 0.01, 0.1, 1],
                         'classifier__kernel': ['rbf'],
                         'scaler': [StandardScaler(), RobustScaler(), None]},
             return_train_score=True, scoring='accuracy')

___

#### 2.1.3 - Visualize Grid Search Results

Print the results from the grid search. The following outputs must be printed.
* Best parameter set
* Best cross-validation train score
* Best cross-validation test score
* Test set accuracy

In [24]:
# TODO: Print the results from the grid search

print("Best params:\n{}\n".format(grid.best_params_))

# Calculate the mean train score for the best parameter set
best_index = grid.best_index_
train_score = grid.cv_results_['mean_train_score'][best_index]

print(f"Best cross-validation train score: {train_score:.2f}")
print(f"Best cross-validation validation score: {grid.best_score_:.2f}")
print(f"Test set accuracy: {grid.score(X_test, y_test):.2f}")


Best params:
{'classifier__C': 100, 'classifier__gamma': 0.01, 'classifier__kernel': 'rbf', 'scaler': RobustScaler()}

Best cross-validation train score: 0.84
Best cross-validation validation score: 0.82
Test set accuracy: 0.83


Display the confusion matrix and print the classification report for the trained model using the *test* split.

In [25]:
# TODO - Display confusion matrix and print classification reports

# Make predictions on the test set
y_pred = grid.predict(X_test)

# Display the confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:\n", conf_matrix)

# Print the classification report
class_report = classification_report(y_test, y_pred)
print("Classification Report:\n", class_report)


Confusion Matrix:
 [[154   1   0]
 [ 20 173  14]
 [  1  44  54]]
Classification Report:
               precision    recall  f1-score   support

        rest       0.88      0.99      0.93       155
    straight       0.79      0.84      0.81       207
        turn       0.79      0.55      0.65        99

    accuracy                           0.83       461
   macro avg       0.82      0.79      0.80       461
weighted avg       0.82      0.83      0.82       461



___

#### 2.1.4 - Question


1. Consider the differences between the performance on the training/cross-validation dataset for section 1 (1.2.3) and section 2 (2.1.3). Which model is a better fit based on *just these two evaluations*? (4 marks)

#### 2.1.4 Answer
1.

- **Performance:** Both models have the same best parameters, cross-validation scores, test-set scores, and confusion matrices.
- **Features:** The model in Section 1 used all the features, while the model in Section 2 used a selected subset of features.

Given that both models achieve the same performance metrics, the model using the selected features (Section 2) could be considered a better fit due to its simplicity and efficiency. It achieves the same high accuracy with fewer features, which is beneficial for computational efficiency and potentially improving model interpretability.



___

### 2.2 - Model Evaluation on Test Datasets 1 and 2 (7 Marks)

___

#### 2.2.1 - Extract selected features and targets from test datasets

The model trained in 2.1 is to be tested on the independent datasets, that is test_set1 and test_set2.
Seperate the testsets into the selected features and target vectors *(X1, y1)* and (X2,)

In [26]:
# TODO - Load the target vector from test_set1 into y1, and the selected features from test_set1 into X1
# (1 mark)
y1 = test_set1['motion_state']
X1 = test_set1[['ax_var', 'ay_var', 'az_sum', 'a_sum', 'gz_var']]

# Display the first two rows of X1
X1.head(2)


ax_var        ay_var    az_sum     a_sum        gz_var
0  0.000001  7.749919e-08  9.786930  9.801661  7.727190e-10
1  0.000001  9.243000e-08  9.788127  9.802481  8.871433e-10

In [27]:
# TODO - Load the target vector from test_set2 into y2, and the selected features from test_set2 into X2
# (1 mark)
y2 = test_set2['motion_state']
X2 = test_set2[['ax_var', 'ay_var', 'az_sum', 'a_sum', 'gz_var']]

# Display the first two rows of X2
X2.head(2)


ax_var        ay_var    az_sum     a_sum        gz_var
0  0.000002  3.020959e-07  9.791451  9.800683  1.405038e-09
1  0.000003  6.859937e-07  9.785619  9.797741  1.164381e-09

___

#### 2.2.2 - Visualize Results

For each test dataset, print the test set accuracy score.

In [28]:
# TODO - Print test score accuracies for test_set1 and test_set2

# Calculate accuracy for test_set1
accuracy1 = grid.score(X1, y1)
print(f"Test set 1 accuracy: {accuracy1:.2f}")

# Calculate accuracy for test_set2
accuracy2 = grid.score(X2, y2)
print(f"Test set 2 accuracy: {accuracy2:.2f}")

Test set 1 accuracy: 0.83
Test set 2 accuracy: 0.84


Display the confusion matrix and print the classification for test_set 1 *(X1, y1)*

In [29]:
# TODO - Display confusion matrix and print classification report for (X1, y1)

# Make predictions on test_set1
y_pred1 = grid.predict(X1)

# Display the confusion matrix for test_set1
conf_matrix1 = confusion_matrix(y1, y_pred1)
print("Confusion Matrix for test_set1:\n", conf_matrix1)

# Print the classification report for test_set1
class_report1 = classification_report(y1, y_pred1)
print("Classification Report for test_set1:\n", class_report1)

Confusion Matrix for test_set1:
 [[229   2   0]
 [ 26 336  49]
 [  0  57  84]]
Classification Report for test_set1:
               precision    recall  f1-score   support

        rest       0.90      0.99      0.94       231
    straight       0.85      0.82      0.83       411
        turn       0.63      0.60      0.61       141

    accuracy                           0.83       783
   macro avg       0.79      0.80      0.80       783
weighted avg       0.83      0.83      0.83       783



Display the confusion matrix and print the classification for test_set 2 *(X2, y2)*

In [30]:
# TODO - Display confusion matrix and print classification report for (X2, y2)

# Make predictions on test_set2
y2_pred = grid.predict(X2)

# Display the confusion matrix
conf_matrix2 = confusion_matrix(y2, y2_pred)
print("Confusion Matrix for test_set2:\n", conf_matrix2)

# Print the classification report
class_report2 = classification_report(y2, y2_pred)
print("Classification Report for test_set2:\n", class_report2)


Confusion Matrix for test_set2:
 [[ 15   1   0]
 [  0 178  28]
 [  0  18  57]]
Classification Report for test_set2:
               precision    recall  f1-score   support

        rest       1.00      0.94      0.97        16
    straight       0.90      0.86      0.88       206
        turn       0.67      0.76      0.71        75

    accuracy                           0.84       297
   macro avg       0.86      0.85      0.85       297
weighted avg       0.85      0.84      0.84       297



___

#### 2.2.3 Questions

1. For section 2, compare between the model's performance on the training/validation (2.1) set to the the performance on the two testing sets (2.2). How does the model fit? (3 marks)
1. Compare the models' performances and fits between section 1 and section 2, and consider all three datasets. (3 marks)

#### 2.2.3 Answers
1.  Comparision between performance on Training/Validation vs Test Sets (Section 2):

The model performed well on the training/validation set with high accuracy (0.98 train, 0.95 validation), but performance dropped on the independent test sets (0.83 and 0.84 accuracy). This suggests overfitting on the training data and good generalization but not perfect on unseen data.

2.Comparison Between Section 1 and Section 2 Models:

* Section 1 (using all features) had better performance on test sets (0.96 accuracy).

*   Section 2 (using selected features) showed slightly lower test accuracy (0.83 and 0.84) but is more efficient and simpler.

*   Section 1 overfits slightly, while Section 2 generalizes better with fewer features.



____

## Section 3 - Classification Using *Group Selected Features* From Inertial Sensor Data (10 Marks)

Once again, an *SVC* model is to be trained and tuned using grid search. However, you must select which features to use from the ones provided in *cv_set*.  

____

### 3.1 Grid Search with Cross Validation (4 Marks)

___

#### 3.1.1 Extract *Group Selected* features and target from *cv_set*

**Question**
1. Please list the features that you have selected and explain *why you have chosen* these features? (4 marks)


**Answer**
1.

The Selected Features are:


*   ax_var
*   ay_var
*   a_sum

Reason for Selection:

The selected features are ax_var, ay_var, and a_sum. ax_var and ay_var capture the variability in movement, essential for identifying motion patterns, while a_sum represents overall movement intensity, helping to distinguish different motion states. These features simplify the model while ensuring high relevance and efficiency. They reduce complexity and noise, improving performance and interpretability.







In [31]:
# TODO - Load group selected features from cv_set to X
X = cv_set[['ax_var', 'ay_var', 'a_sum']]
# Display the first two rows of X
X.head(2)

ax_var        ay_var     a_sum
0  8.046094e-07  1.319665e-07  9.797418
1  1.032619e-06  1.073067e-07  9.798815

In [32]:
# TODO - Load motion_state column (target vector) into y
y = cv_set['motion_state']

# Display the first two rows of y
y.head(2)

0    rest
1    rest
Name: motion_state, dtype: object

___

#### 3.1.2 - Configure and Perform Grid Search with Cross-Validation

Follow the steps from **1.2.2** and perform model training and tuning using the group selected feature set.

In [33]:
# TODO - Perform model training and tuning using `GridSearchCV`

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42, stratify=y)

# Select the same features for X_test to match the training data
X_test = X_test[['ax_var', 'ay_var', 'a_sum']]

# Define the pipeline with a scaler and an SVC classifier
pipe = Pipeline([
    ('scaler', StandardScaler()),
    ('classifier', SVC())
])

# Setup grid for hyperparameter tuning
param_grid = {
    'scaler': [StandardScaler(), RobustScaler(), None],  # Scalers to test
    'classifier__C': [0.01, 0.1, 1, 10, 100],  # Regularization parameter
    'classifier__gamma': [0.001, 0.01, 0.1, 1],  # Kernel coefficient
    'classifier__kernel': ['rbf']  # Only using rbf kernel for SVC
}

# Perform GridSearchCV with cross-validation
grid = GridSearchCV(pipe, param_grid, cv=5, scoring='accuracy', return_train_score=True)
grid.fit(X, y)


GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('scaler', StandardScaler()),
                                       ('classifier', SVC())]),
             param_grid={'classifier__C': [0.01, 0.1, 1, 10, 100],
                         'classifier__gamma': [0.001, 0.01, 0.1, 1],
                         'classifier__kernel': ['rbf'],
                         'scaler': [StandardScaler(), RobustScaler(), None]},
             return_train_score=True, scoring='accuracy')

___

#### 3.1.3 - Visualize Grid Search Results

Print the results from the grid search. The following outputs must be printed.
* Best parameter set
* Best cross-validation train score
* Best cross-validation test score
* Test set accuracy

In [34]:
# TODO: Print the results from the grid search


# Best parameter set
print("Best params:\n{}".format(grid.best_params_))

# Best cross-validation train score
best_index = grid.best_index_
train_score = grid.cv_results_['mean_train_score'][best_index]
print("Best cross-validation train score: {:.2f}".format(train_score))

# Best cross-validation test score
print("Best cross-validation validation score: {:.2f}".format(grid.best_score_))

# Test-set score (evaluating the model with best parameters on test data)
test_score = grid.score(X_test, y_test)
print("Test-set score: {:.2f}".format(test_score))

Best params:
{'classifier__C': 100, 'classifier__gamma': 0.1, 'classifier__kernel': 'rbf', 'scaler': RobustScaler()}
Best cross-validation train score: 0.78
Best cross-validation validation score: 0.75
Test-set score: 0.78


Display the confusion matrix and print the classification report for the trained model using the *test* split.

In [35]:
# TODO - Display confusion matrix and print classification report

# Make predictions on the test set
y_pred = grid.predict(X_test)

# Display the confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:\n", conf_matrix)

# Print the classification report
class_report = classification_report(y_test, y_pred)
print("Classification Report:\n", class_report)

Confusion Matrix:
 [[151   4   0]
 [ 20 184   3]
 [  4  71  24]]
Classification Report:
               precision    recall  f1-score   support

        rest       0.86      0.97      0.92       155
    straight       0.71      0.89      0.79       207
        turn       0.89      0.24      0.38        99

    accuracy                           0.78       461
   macro avg       0.82      0.70      0.70       461
weighted avg       0.80      0.78      0.74       461



___

### 3.2 - Model Evaluation on Test Datasets 1 and 2 (6 Marks)

___

#### 3.2.1 - Extract group selected features and targets from test datasets

The model trained in 3.1 is to be tested on the independent datasets, that is test_set1 and test_set2.
Seperate the testsets into group selected features and target vectors *(X1, y1)* and *(X2,y2)*

In [36]:
# TODO - Load the target vector from test_set1 into y1, and the group selected features from test_set1 into X1
y1 = test_set1['motion_state']
X1 = test_set1[['ax_var', 'ay_var', 'a_sum']]
X1.head(2)

ax_var        ay_var     a_sum
0  0.000001  7.749919e-08  9.801661
1  0.000001  9.243000e-08  9.802481

In [37]:
# TODO - Load the target vector from test_set2 into y2, and the group selected features from test_set2 into X2
y2 = test_set2['motion_state']
X2 = test_set2[['ax_var', 'ay_var', 'a_sum']]
X2.head(2)

ax_var        ay_var     a_sum
0  0.000002  3.020959e-07  9.800683
1  0.000003  6.859937e-07  9.797741

___

#### 3.2.2 - Visualize Results

For each test dataset, print the test set accuracy score.

In [38]:
# TODO - Print test score accuracies for test_set1 and test_set2

# Calculate the test set accuracy for test_set1
test_set1_score = grid.score(X1, y1)
print("test_set1 score: {:.2f}".format(test_set1_score))

# Calculate the test set accuracy for test_set2
test_set2_score = grid.score(X2, y2)
print("test_set2 score: {:.2f}".format(test_set2_score))

test_set1 score: 0.79
test_set2 score: 0.67


Display the confusion matrix and print the classification for test_set 1 *(X1, y1)*

In [39]:
# TODO - Display confusion matrix and print classification report for (X1, y1)

# Make predictions for test_set1
y1_pred = grid.predict(X1)

# Display the confusion matrix
conf_matrix_1 = confusion_matrix(y1, y1_pred)
print("Confusion Matrix for Test Set 1:\n", conf_matrix_1)

# Print the classification report
class_report_1 = classification_report(y1, y1_pred)
print("Classification Report for Test Set 1:\n", class_report_1)

Confusion Matrix for Test Set 1:
 [[228   3   0]
 [ 16 351  44]
 [  0 104  37]]
Classification Report for Test Set 1:
               precision    recall  f1-score   support

        rest       0.93      0.99      0.96       231
    straight       0.77      0.85      0.81       411
        turn       0.46      0.26      0.33       141

    accuracy                           0.79       783
   macro avg       0.72      0.70      0.70       783
weighted avg       0.76      0.79      0.77       783



Display the confusion matrix and print the classification for test_set 2 *(X2, y2)*

In [40]:
# TODO - Display confusion matrix and print classification report for (X2, y2)

# Make predictions for test_set2
y2_pred = grid.predict(X2)

# Display the confusion matrix for test_set2
conf_matrix2 = confusion_matrix(y2, y2_pred)
print("Confusion Matrix for Test Set 2:")
print(conf_matrix2)

# Print the classification report for test_set2
class_report2 = classification_report(y2, y2_pred)
print("Classification Report for Test Set 2:")
print(class_report2)


Confusion Matrix for Test Set 2:
[[ 13   3   0]
 [  1 158  47]
 [  1  45  29]]
Classification Report for Test Set 2:
              precision    recall  f1-score   support

        rest       0.87      0.81      0.84        16
    straight       0.77      0.77      0.77       206
        turn       0.38      0.39      0.38        75

    accuracy                           0.67       297
   macro avg       0.67      0.66      0.66       297
weighted avg       0.68      0.67      0.67       297



___

#### 3.2.3 Questions

1. Based on the observations across all three datasets, compare the performance of the model trained in this section to: (3 marks)
    1. Model trained in Section 1 using *all features*.
    1. Model trained in Section 2 using *selected features*.

2. Please proivde potential explanations for the differences observed in models' performances across the three sections. (3 marks)

#### 3.2.3 Answers
1.
 Compare the model's performance:


*   Model trained in Section 1 (all features):

test_set1: Accuracy = 0.78 (same as Section 3).

test_set2: Accuracy = 0.75 (better than Section 3, which is 0.70).

*   Model trained in Section 2 (selected features):

test_set1: Accuracy = 0.76 (slightly lower than Section 3).

test_set2: Accuracy = 0.69 (similar to Section 3, which is 0.70).



  

2. Explanations for differences in performance:

a. Feature selection: Selected features in Section 3 improved performance on test_set1 but reduced accuracy on test_set2 compared to using all features.

b. Hyperparameter tuning: Grid search in Section 3 helped improve the model’s performance on test_set1.

c. Dataset variability: test_set2 had lower accuracy due to class imbalance and complexity in distinguishing "turn" movements.

d. Data scaling: The RobustScaler in Section 3 helped improve results on test_set1 but wasn't as effective for test_set2.


____

## Section 4 - Classification Using Principal Components of Features From Inertial Sensor Data (10 Marks)

____

### 4.1 Principal Component Analysis (PCA) Search with Cross Validation (4 Marks)

___

#### 4.1.1 Extract *ALL* features and target from *cv_set*

In [41]:
# TODO - Load all features from cv_set to X
X = cv_set.drop('motion_state', axis=1)
X.head(2)

ax_sum    ay_sum    az_sum    gx_sum    gy_sum    gz_sum     a_sum  \
0 -0.003929 -0.093081  9.789349 -0.002649  0.001406 -0.001391  9.797418   
1 -0.000966 -0.092417  9.789095 -0.001806  0.000103 -0.001943  9.798815   

      g_sum        ax_var        ay_var  ...     a_min     g_min    ax_max  \
0  0.022554  8.046094e-07  1.319665e-07  ...  0.023871  0.000003  0.001496   
1  0.021731  1.032619e-06  1.073067e-07  ...  0.024153  0.000006  0.001865   

     ay_max    az_max    gx_max    gy_max    gz_max     a_max     g_max  
0  0.000544  0.024992  0.000110  0.000075  0.000078  0.024997  0.000187  
1  0.000615  0.024828  0.000124  0.000098  0.000090  0.024829  0.000144  

[2 rows x 32 columns]

In [42]:
# TODO - Load motion_state column (target vector) into y
y = cv_set['motion_state']

___

#### 4.1.2 - Configure and Perform Grid Search with Cross-Validation

Similar to the steps from **1.2.2** and perform model training and tuning for an `SVC` model. Perform the `train_test_split`, initiate a `Pipeline`, then define the *Parameters* for `GridSearchCV`. Finally, perfrom the grid search.

In [43]:
# TODO - Perform model training and tuning using `GridSearchCV`
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42, stratify=y)

In the `Pipeline` add an entry *'pca'* to apply `PCA`.

In [44]:
# Perform any missing imports
from sklearn.decomposition import PCA
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, RobustScaler
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split, GridSearchCV

# TODO - Initiate the pipeline
# (1 marks)
pipe = Pipeline([
    ('scaler', StandardScaler()),  # StandardScaler to normalize data
    ('pca', PCA()),                # Principal Component Analysis (PCA) step
    ('classifier', SVC())          # Support Vector Classifier (SVC)
])

Similar to previous sections, define the grid search configuration (*param_grid*). Add *n_components* hyper parameter tuning for PCA, and search over the values `n_components=[4,7,11,14,16]`.  

In [45]:
# TODO - Define the parameter grid for your model
# (3 marks)
param_grid = {
    'scaler': [StandardScaler(), RobustScaler(), None],  # Scalers to test
    'pca__n_components': [4, 7, 11, 14, 16],  # Number of PCA components to test
    'classifier__C': [0.01, 0.1, 1, 10, 100],  # Regularization parameter for SVC
    'classifier__gamma': [0.001, 0.01, 0.1, 1],  # Kernel coefficient for SVC
    'classifier__kernel': ['rbf']  # Using the rbf kernel for SVC
}

In [46]:
# Perform the grid search to train model and tune hyper parameters
grid = GridSearchCV(pipe, param_grid, cv=5, scoring='accuracy', return_train_score=True)
grid.fit(X_train, y_train)

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('scaler', StandardScaler()),
                                       ('pca', PCA()), ('classifier', SVC())]),
             param_grid={'classifier__C': [0.01, 0.1, 1, 10, 100],
                         'classifier__gamma': [0.001, 0.01, 0.1, 1],
                         'classifier__kernel': ['rbf'],
                         'pca__n_components': [4, 7, 11, 14, 16],
                         'scaler': [StandardScaler(), RobustScaler(), None]},
             return_train_score=True, scoring='accuracy')

___

#### 4.1.3 - Visualize Grid Search Results

Print the results from the grid search. The following outputs must be printed.
* Best parameter set
* Best cross-validation train score
* Best cross-validation test score
* Test set accuracy

In [47]:
# TODO: Print the results from the grid search

# Best parameter set
print("Best params:\n{}".format(grid.best_params_))

# Best cross-validation train score
best_index = grid.best_index_
train_score = grid.cv_results_['mean_train_score'][best_index]
print("Best cross-validation train score: {:.2f}".format(train_score))

# Best cross-validation test score
print("Best cross-validation validation score: {:.2f}".format(grid.best_score_))

# Test-set score (evaluating the model with best parameters on test data)
test_score = grid.score(X_test, y_test)
print("Test-set score: {:.2f}".format(test_score))


Best params:
{'classifier__C': 100, 'classifier__gamma': 0.01, 'classifier__kernel': 'rbf', 'pca__n_components': 14, 'scaler': RobustScaler()}
Best cross-validation train score: 0.97
Best cross-validation validation score: 0.94
Test-set score: 0.96


Display the confusion matrix and print the classification report for the trained model using the *test* split.

In [48]:
# TODO - Display confusion matrix and print classification report
y_pred = grid.predict(X_test)

# Display the confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:\n", conf_matrix)

# Print the classification report
class_report = classification_report(y_test, y_pred)
print("Classification Report:\n", class_report)


Confusion Matrix:
 [[154   1   0]
 [  9 192   6]
 [  0   2  97]]
Classification Report:
               precision    recall  f1-score   support

        rest       0.94      0.99      0.97       155
    straight       0.98      0.93      0.96       207
        turn       0.94      0.98      0.96        99

    accuracy                           0.96       461
   macro avg       0.96      0.97      0.96       461
weighted avg       0.96      0.96      0.96       461



___

### 4.2 - Model Evaluation on Test Datasets 1 and 2 (6 Marks)

___

#### 4.2.1 - Extract group selected features and targets from test datasets

In [49]:
# TODO - Load the target vector from test_set1 into y1, and all features from test_set1 into X1
y1 = test_set1['motion_state']
X1 = test_set1.drop(columns=['motion_state'])
X1.head(2)

ax_sum    ay_sum    az_sum    gx_sum    gy_sum    gz_sum     a_sum  \
0  0.264465  0.089904  9.786930 -0.002305  0.000218 -0.001143  9.801661   
1  0.264300  0.088882  9.788127 -0.001865 -0.000523 -0.001999  9.802481   

      g_sum    ax_var        ay_var  ...     a_min     g_min    ax_max  \
0  0.025634  0.000001  7.749919e-08  ...  0.024056  0.000003  0.002506   
1  0.025218  0.000001  9.243000e-08  ...  0.024084  0.000007  0.002427   

     ay_max    az_max    gx_max    gy_max    gz_max     a_max     g_max  
0  0.000845  0.024884  0.000155  0.000078  0.000069  0.024886  0.000170  
1  0.000949  0.024881  0.000201  0.000095  0.000076  0.024883  0.000201  

[2 rows x 32 columns]

In [50]:
# TODO - Load the target vector from test_set2 into y2, and all features from test_set2 into X2
y2 = test_set2['motion_state']
X2 = test_set2.drop(columns=['motion_state'])
X2.head(2)

ax_sum    ay_sum    az_sum    gx_sum    gy_sum    gz_sum     a_sum  \
0 -0.046400  0.305907  9.791451 -0.002184  0.001311 -0.001901  9.800683   
1 -0.050073  0.311947  9.785619 -0.003369  0.001898 -0.002051  9.797741   

      g_sum    ax_var        ay_var  ...     a_min     g_min    ax_max  \
0  0.012437  0.000002  3.020959e-07  ...  0.048406  0.000007  0.002575   
1  0.013805  0.000003  6.859937e-07  ...  0.047696  0.000013  0.003880   

     ay_max    az_max    gx_max    gy_max    gz_max     a_max     g_max  
0  0.002993  0.049701  0.000094  0.000115  0.000080  0.049764  0.000142  
1  0.004106  0.050477  0.000140  0.000140  0.000083  0.050763  0.000151  

[2 rows x 32 columns]

___

#### 4.2.2 - Visualize Results

In [51]:
# TODO - Print test score accuracies for test_set1 and test_set2
print("test_set1 score: {:.2f}".format(grid.score(X1, y1)))
print("test_set2 score: {:.2f}".format(grid.score(X2, y2)))

test_set1 score: 0.93
test_set2 score: 0.67


In [52]:
# TODO - Display confusion matrix and print classification report for (X1, y1)

# Make predictions on test_set1 (X1)
y1_pred = grid.predict(X1)

# Display the confusion matrix for test_set1
conf_matrix1 = confusion_matrix(y1, y1_pred)
print("Confusion Matrix for Test Set 1:\n", conf_matrix1)

# Print the classification report for test_set1
class_report1 = classification_report(y1, y1_pred)
print("Classification Report for Test Set 1:\n", class_report1)


Confusion Matrix for Test Set 1:
 [[223   8   0]
 [ 20 372  19]
 [  0   6 135]]
Classification Report for Test Set 1:
               precision    recall  f1-score   support

        rest       0.92      0.97      0.94       231
    straight       0.96      0.91      0.93       411
        turn       0.88      0.96      0.92       141

    accuracy                           0.93       783
   macro avg       0.92      0.94      0.93       783
weighted avg       0.93      0.93      0.93       783



Display the confusion matrix and print the classification for test_set 2 *(X2, y2)*

In [53]:
# TODO - Display confusion matrix and print classification report for (X2, y2)

# Predict on test_set2
y2_pred = grid.predict(X2)

# Confusion matrix for test_set2
conf_matrix_test2 = confusion_matrix(y2, y2_pred)
print("Confusion Matrix for Test Set 2:")
print(conf_matrix_test2)

# Classification report for test_set2
class_report_test2 = classification_report(y2, y2_pred)
print("Classification Report for Test Set 2:")
print(class_report_test2)


Confusion Matrix for Test Set 2:
[[  5  11   0]
 [  0 119  87]
 [  0   1  74]]
Classification Report for Test Set 2:
              precision    recall  f1-score   support

        rest       1.00      0.31      0.48        16
    straight       0.91      0.58      0.71       206
        turn       0.46      0.99      0.63        75

    accuracy                           0.67       297
   macro avg       0.79      0.63      0.60       297
weighted avg       0.80      0.67      0.67       297



___

#### 4.2.3 - Questions

1. Based on the observations across all three datasets, compare the performance of the model with PCA, trained in this section to: (4 marks)
    * Model trained in Section 1 using *all features*.
    * Model trained in Section 2 using *selected features*.
    * Model trained in Section 3 using *group selected features*

2. Please provide potential explanations for the differences observed in models' performances across between section 1 and 4. (2 marks)

#### 4.2.3 - Answers

Comparison of Model Performance with PCA:

*   Section 1 (All Features):

a. Test set 1 accuracy: 0.93, test set 2 accuracy: 0.67.

b. PCA in Section 4 gave similar performance on test set 1 but lower on test set 2.



*  Section 2 (Selected Features):

a. Moderate accuracy on both sets.

b. PCA slightly reduced performance on test set 1 and more on test set 2.

*   Section 3 (Group Selected Features):

a. Similar accuracy to Section 1 on test set 1 but lower on test set 2.

b. PCA had a slight decrease in performance, especially on test set 2.


Explanations for Differences Between Section 1 and Section 4:

a. PCA Reduces Variance: PCA compresses features, losing some information, leading to reduced performance on complex test sets like test_set2.

b. Loss of Information: In Section 1, all features were used, providing more complete data for generalization compared to PCA’s transformed features.

____

## Section 5 - Classification Using *Pre Selected Features* Derived From Inertial Sensor Data (10 Marks)

___

Now, a second classifer is to be trained and tuned using grid search. The **selected features from section 2** must be used. You may choose from one of the algorithms discussed that class listed below.
1. Random Forest
1. K-Nearest Neighbors
1. Gradient Boosting Machines
1. Neural Networks

____

### 5.1 Grid Search with Cross Validation (6 Marks)

___

#### 5.1.1 Extract *Selected* features and target from *cv_set*

**Questions**
1. Please list the ML model that you have selected . And explain *why you have chosen* this model? (2 marks)

**Answer**
1.  
Model: Random Forest

 Reason: Random Forest is a reliable choice because it handles both categorical and continuous data well and avoids overfitting by combining multiple decision trees. It’s great for capturing complex patterns without much preprocessing and even highlights which features are most important. Plus, it’s scalable, robust to noise, and works well on new data.

In [54]:
# TODO - Load group selected features from cv_set to X

selected_features = ['ax_var', 'ay_var', 'az_sum', 'a_sum', 'gz_var']
X = cv_set[selected_features]
X.head(2)

ax_var        ay_var    az_sum     a_sum        gz_var
0  8.046094e-07  1.319665e-07  9.789349  9.797418  7.446916e-10
1  1.032619e-06  1.073067e-07  9.789095  9.798815  6.983481e-10

In [55]:
# TODO - Load motion_state column (target vector) into y
y = cv_set['motion_state']

___

#### 5.1.2 - Configure and Perform Grid Search with Cross-Validation

Similar to the steps from **1.2.2** and perform model training and tuning using for the model selected. Perform the `train_test_split`, initiate a `Pipeline`, then define the *Parameters* for `GridSearchCV`. Finally, perfrom the grid search.

In [56]:
# TODO - Perform train/test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [57]:
# Perform any missing imports
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
# TODO - Initiate the pipeline
# (1 mark)
pipe = Pipeline([
    ('scaler', StandardScaler()),            # Scaling features
    ('classifier', RandomForestClassifier())  # Using Random Forest classifier
])

In [58]:
# TODO - Define the parameter grid for your model
# (3 marks)
param_grid = {
    'classifier__n_estimators': [100, 200, 300],  # Number of trees in the forest
    'classifier__max_depth': [None, 10, 20],  # Maximum depth of the trees
    'classifier__min_samples_split': [2, 5, 10],  # Minimum number of samples required to split an internal node
    'classifier__min_samples_leaf': [1, 2, 4],  # Minimum number of samples required to be at a leaf node
    'scaler': [StandardScaler(), RobustScaler(), None]  # Scalers to test
}

In [59]:
# TODO - Perform the grid search
grid = GridSearchCV(pipe, param_grid, cv=5, scoring='accuracy', return_train_score=True)
grid.fit(X_train, y_train)

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('scaler', StandardScaler()),
                                       ('classifier',
                                        RandomForestClassifier())]),
             param_grid={'classifier__max_depth': [None, 10, 20],
                         'classifier__min_samples_leaf': [1, 2, 4],
                         'classifier__min_samples_split': [2, 5, 10],
                         'classifier__n_estimators': [100, 200, 300],
                         'scaler': [StandardScaler(), RobustScaler(), None]},
             return_train_score=True, scoring='accuracy')

___

#### 5.1.3 - Visualize Grid Search Results

Print the results from the grid search. The following outputs must be printed.
* Best parameter set
* Best cross-validation train score
* Best cross-validation test score
* Test set accuracy

In [60]:
# TODO: Print the results from the grid search

print("Best params:\n{}".format(grid.best_params_))
print("Best cross-validation train score: {:.2f}".format(grid.best_score_))
print("Best cross-validation validation score: {:.2f}".format(grid.score(X_test, y_test)))
print("Test-set score: {:.2f}".format(grid.score(X_test, y_test)))

Best params:
{'classifier__max_depth': 10, 'classifier__min_samples_leaf': 1, 'classifier__min_samples_split': 10, 'classifier__n_estimators': 100, 'scaler': StandardScaler()}
Best cross-validation train score: 0.85
Best cross-validation validation score: 0.85
Test-set score: 0.85


Display the confusion matrix and print the classification report for the trained model using the *test* split.

In [61]:
# TODO - Display confusion matrix and print classification report
y_pred = grid.predict(X_test)

# Display the confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:\n", conf_matrix)

# Print classification report
class_report = classification_report(y_test, y_pred)
print("Classification Report:\n", class_report)

Confusion Matrix:
 [[102   2   0]
 [  5 120  12]
 [  0  27  39]]
Classification Report:
               precision    recall  f1-score   support

        rest       0.95      0.98      0.97       104
    straight       0.81      0.88      0.84       137
        turn       0.76      0.59      0.67        66

    accuracy                           0.85       307
   macro avg       0.84      0.82      0.82       307
weighted avg       0.85      0.85      0.85       307



___

### 5.2 - Model Evaluation on Test Datasets 1 and 2 (4 Marks)

___

#### 5.2.1 - Extract selected features and targets from test datasets

The model trained in 5.1, is to be tested on the independent datasets, that is test_set1 and test_set2.
Seperate the testsets into features and target vectors *(X1, y1)* and *(X2,y2)*

In [62]:
# TODO - Load the target vector from test_set1 into y1, and the selected features from test_set1 into X1
y1 = test_set1['motion_state']
X1 = test_set1[selected_features]
X1.head(2)

ax_var        ay_var    az_sum     a_sum        gz_var
0  0.000001  7.749919e-08  9.786930  9.801661  7.727190e-10
1  0.000001  9.243000e-08  9.788127  9.802481  8.871433e-10

In [63]:
# TODO - Load the target vector from test_set2 into y2, and the selected features from test_set2 into X2
y2 = test_set2['motion_state']
X2 = test_set2[selected_features]
X2.head(2)

ax_var        ay_var    az_sum     a_sum        gz_var
0  0.000002  3.020959e-07  9.791451  9.800683  1.405038e-09
1  0.000003  6.859937e-07  9.785619  9.797741  1.164381e-09

___

#### 5.2.2 - Visualize Results

For each test dataset, print the test set accuracy score.

In [64]:
# TODO - Print test score accuracies for test_set1 and test_set2

# Test accuracy for test_set1
test_set1_score = grid.score(X1, y1)
print("test_set1 score: {:.2f}".format(test_set1_score))

# Test accuracy for test_set2
test_set2_score = grid.score(X2, y2)
print("test_set2 score: {:.2f}".format(test_set2_score))

test_set1 score: 0.85
test_set2 score: 0.66


Display the confusion matrix and print the classification for test_set 1 *(X1, y1)*

In [65]:
# TODO - Display confusion matrix and print classification report for (X1, y1)

# Predict for test_set1
y_pred1 = grid.predict(X1)

# Display the confusion matrix
conf_matrix1 = confusion_matrix(y1, y_pred1)
print("Confusion Matrix for Test Set 1:\n", conf_matrix1)

# Print the classification report
class_report1 = classification_report(y1, y_pred1)
print("Classification Report for Test Set 1:\n", class_report1)



Confusion Matrix for Test Set 1:
 [[224   7   0]
 [  4 359  48]
 [  0  59  82]]
Classification Report for Test Set 1:
               precision    recall  f1-score   support

        rest       0.98      0.97      0.98       231
    straight       0.84      0.87      0.86       411
        turn       0.63      0.58      0.61       141

    accuracy                           0.85       783
   macro avg       0.82      0.81      0.81       783
weighted avg       0.85      0.85      0.85       783



Display the confusion matrix and print the classification for test_set 2 *(X2, y2)*

In [66]:
# TODO - Display confusion matrix and print classification report for (X2, y2)

# TODO - Display confusion matrix and print classification report for (X2, y2)
y_pred2 = grid.predict(X2)

# Display the confusion matrix
conf_matrix2 = confusion_matrix(y2, y_pred2)
print("Confusion Matrix for Test Set 2:\n", conf_matrix2)

# Print classification report
class_report2 = classification_report(y2, y_pred2)
print("Classification Report for Test Set 2:\n", class_report2)


Confusion Matrix for Test Set 2:
 [[  1  15   0]
 [  0 125  81]
 [  0   4  71]]
Classification Report for Test Set 2:
               precision    recall  f1-score   support

        rest       1.00      0.06      0.12        16
    straight       0.87      0.61      0.71       206
        turn       0.47      0.95      0.63        75

    accuracy                           0.66       297
   macro avg       0.78      0.54      0.49       297
weighted avg       0.77      0.66      0.66       297



___

#### 5.2.3 Question

1. For the model trainied in this section, compare its performance to the performance of SVC which was also trained with the same selected features (section 2). Which model fits better? Explain causes for any observed discrepancies? (4 marks)

#### 5.2.3 - Answer
1.

The Random Forest model outperforms the SVC on test set 1, achieving higher accuracy (0.85 vs. 0.81). This advantage comes from its ensemble learning approach, which effectively handles class imbalances and captures complex feature interactions.

However, on test set 2, the situation reverses, with Random Forest performing worse (0.67 vs. 0.73 for SVC). This difference may stem from the dataset’s complexity and feature structure. While Random Forest can struggle with noisy or highly complex data, SVC’s sensitivity to feature scaling and structure allows it to adapt better in this case, though it depends heavily on proper tuning to avoid overfitting or underfitting.


____